src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Configuration/ConfigurationProjectConfigurationDimensionProviderTest.cs

```diff
@@ -0,0 +1,189 @@
+                Assert.Equal("Debug;Release;CustomConfiguration", property.Value);
+
+                // On ChangeEventStage.Before the property should be added
+                args = new ProjectConfigurationDimensionValueChangedEventArgs(
+                    unconfiguredProject,
+                    ConfigurationDimensionChange.Add,
+                    ChangeEventStage.Before,
+                    ConfigurationGeneral.ConfigurationProperty,
+                    "CustomConfig");
+                await provider.OnDimensionValueChangedAsync(args);
+                property = MsBuildUtilities.GetProperty(projectFile.Project, Configurations);
+                Assert.NotNull(property);
+                Assert.Equal("Debug;Release;CustomConfiguration;CustomConfig", property.Value);
+            }
+        }
+
+        [Fact]
+        public async void ConfigurationProjectConfigurationDimensionProvider_OnDimensionValueChanged_Remove()
+        {
+            using (var projectFile = new MsBuildTempProjectFile(projectXml))
+            {
+                IProjectXmlAccessor _projectXmlAccessor = IProjectXmlAccessorFactory.Create(projectFile.Project);
+                var provider = new ConfigurationProjectConfigurationDimensionProvider(_projectXmlAccessor);
+                var unconfiguredProject = UnconfiguredProjectFactory.Create(filePath: projectFile.Filename);
+
+                // On ChangeEventStage.After nothing should be changed
+                ProjectConfigurationDimensionValueChangedEventArgs args = new ProjectConfigurationDimensionValueChangedEventArgs(
+                    unconfiguredProject,
+                    ConfigurationDimensionChange.Delete,
+                    ChangeEventStage.After,
+                    ConfigurationGeneral.ConfigurationProperty,
+                    "CustomConfiguration");
+                await provider.OnDimensionValueChangedAsync(args);
+                var property = MsBuildUtilities.GetProperty(projectFile.Project, Configurations);
+                Assert.NotNull(property);
+                Assert.Equal("Debug;Release;CustomConfiguration", property.Value);
+
+                // On ChangeEventStage.Before the property should be removed
+                args = new ProjectConfigurationDimensionValueChangedEventArgs(
+                    unconfiguredProject,
+                    ConfigurationDimensionChange.Delete,
+                    ChangeEventStage.Before,
+                    ConfigurationGeneral.ConfigurationProperty,
+                    "CustomConfiguration");
+                await provider.OnDimensionValueChangedAsync(args);
+                property = MsBuildUtilities.GetProperty(projectFile.Project, Configurations);
+                Assert.NotNull(property);
+                Assert.Equal("Debug;Release", property.Value);
+            }
+        }
+
+        [Fact]
+        public async void ConfigurationProjectConfigurationDimensionProvider_OnDimensionValueChanged_Rename()
```

Comment:
I didn't add it (or other negative tests for remove/rename) because those operations are currently a no-op and from a full integration point of view you can't get into that state (the provider is the source of truth so the data would have to get corrupted on the consumer before the request to rename comes back). That being said, if we do change the behavior of the utilities functions I'll go ahead and add tests for that.